pkg/api/resource/quantity.go

```diff
@@ -0,0 +1,402 @@
+// <sign>            ::= "+" | "-"
+// <digits>          ::= <digit> | <digit><digits>
+// <digit>           ::= 0 | 1 | ... | 9
+// <exponent>        ::= <binarySuffix> | <decimalExponent> | <decimalSuffix>
+// <binarySuffix>    ::= i | Ki | Mi | Gi | Ti | Pi | Ei
+// <decimalSuffix>   ::= m | "" | k | M | G | T | P | E
+// <decimalExponent> ::= "e" <digits> | "E" <digits>
+//   (Where digits is always a multiple of 3)
+//   (Note that 1024 = 1Ki but 1000 = 1k; I didn't choose the capitalization.)
+//
+// No matter which of the three exponent forms is used, no quantity may represent
+// a number less than 1m or greater than 2^63-1 in magnitude. Numbers that exceed
+// a bound will be capped at that bound. (E.g.: 0.1m will be treated as 1m.)
+// This may be extended in the future if we require larger or smaller quantities.
+//
+// Numbers with binary suffixes may not have any fractional part.
+//
+// Quantities will be serialized in the same format that they were parsed from.
+// Before serializing, Quantity will be put in "canonical form".
+// This means that Exponent will be adjusted up or down (with a
+// corresponding increase or decrease in Mantissa) until one of the
+// following is true:
+//   a. Binary SI mode: Mantissa mod 1024 is nonzero.
+//        Examples: 1Gi 300Mi 6Ki 1001Gi
+//        Non-canonical: 1024Gi (Should be 1Ti)
+//   b. Decimal SI mode: exponent is greater than 3 and one of Mantissa's three least
+//      significant digits is nonzero.
+//        Examples: 1G 300M 3K 1001G
+//        Non-canonical: 1000G (Should be 1T)
+//   c. Decimal SI mode: exponent is less than or equal to zero, and Mantissa has no more
+//      than three nonzero decimals. If any decimals are nonzero, three
+//      decimals will be emitted.
+//        Examples: 5 123.450 1.001 0.045
+//        Non-canonical: 1m (should be 0.001)
+//   d. Decimal exponent mode: as for decimal SI mode, but using the corresponding
+//      "e6" or "E6" form.
+//
+// The sign will be omitted unless the number is negative.
+//
+// Note that the quantity will NEVER be represented by a floating point number. That is
+// the whole point of this exercise.
+//
+// Non-canonical values will still parse as long as they are well formed,
+// but will be re-emitted in their canonical form. (So always use canonical
+// form, or don't diff.)
+//
+// This format is intended to make it difficult to use these numbers without
+// writing some sort of special handling code in the hopes that that will
+// cause implementors to also use a fixed point implementation.
+type Quantity struct {
+	// Amount is public, so you can manipulate it if the accessor
+	// functions are not sufficient.
+	Amount *inf.Dec
+
+	// Change Format at will. Note that when set to BinarySI, fractional
+	// parts will be rounded off by String().
+	Format
+}
+
+const (
+	splitREString = "^([+-]?[0123456789.]+)([eEimkKMGTP]*[-+]?[0123456789]*)$"
```

Comment:
This is more general than actually permitted.  Maybe add a comment to say that this is OK, and conformance will be checked later.